File: ObjC/ObjCTests/PointTests.m
//
//  PointTests.m
//  ObjC
//
//  Created by harry on 2/28/15.
//  Copyright (c) 2015 Harry Cheung. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "HCMPoint.h"

@interface PointTests : XCTestCase

@end

@implementation PointTests

- (void)testSubtraction {
  HCMPoint *a =[[HCMPoint alloc] initWithLatitude:5 longitude:5];
  HCMPoint *b =[[HCMPoint alloc] initWithLatitude:10 longitude:10];
  HCMPoint *c = [b subtractWithPoint:a];
  
  XCTAssertEqual([c latitudeDegrees], 5.0);
  XCTAssertEqual([c longitudeDegrees], 5.0);
}

- (void)testBearingTo {
  HCMPoint *a =[[HCMPoint alloc] initWithLatitude:5 longitude:5];
  HCMPoint *b =[[HCMPoint alloc] initWithLatitude:5 longitude:10];
  
  XCTAssertEqual([a bearingToPoint:b], 89.781973);
  XCTAssertEqual([a bearingToPoint:b inRadians: true], 1.566991);
}

- (void)testDestination {
  HCMPoint *a = [[HCMPoint alloc] initWithLatitude:37.452602 longitude:-122.207069];
  HCMPoint *d = [a destinationWithBearing:308 distance:50];
  
  XCTAssertEqual([d latitudeDegrees], 37.452879);
  XCTAssertEqual([d longitudeDegrees], -122.207515);
}

- (void)testDistanceTo {
  HCMPoint *a =[[HCMPoint alloc] initWithLatitude:50.06639 longitude: -5.71472];
  HCMPoint *b =[[HCMPoint alloc] initWithLatitude:58.64389 longitude: -3.07000];
  
  XCTAssertEqualWithAccuracy([a distanceToPoint:b], 968853.52, 0.001);
}

- (void)testIntersectSimple {
  HCMPoint *a =[[HCMPoint alloc] initWithLatitude:5 longitude:5];
  HCMPoint *b =[[HCMPoint alloc] initWithLatitude:15 longitude:15];
  HCMPoint *c =[[HCMPoint alloc] initWithLatitude:5 longitude:15];
  HCMPoint *d =[[HCMPoint alloc] initWithLatitude:15 longitude:5];
  
  HCMPoint *i = [HCMPoint intersectSimpleWithPoint:a
                                         withPoint:b
                                         withPoint:c
                                         withPoint:d];
  
  XCTAssertEqual([i latitudeDegrees], 10);
  XCTAssertEqual([i longitudeDegrees], 10);
}

@end

File: SwiftApp/SwiftAppTests/PointTests.swift
//
//  Copyright (c) 2015 Harry Cheung
//

import UIKit
import XCTest

class PointTests: XCTestCase {
  
  func testEquals() {
    let a = Point(latitude: 14, longitude: 14)
    let b = Point(latitude: 14, longitude: 14)
    let c = Point(latitude: 25, longitude: 25)
    
    XCTAssertEqual(a, b)
    XCTAssertNotEqual(a, c)
  }
  
  func testSubtraction() {
    let a = Point(latitude: 5, longitude: 5)
    let b = Point(latitude: 10, longitude: 10)
    let c = b.subtract(a)
    
    XCTAssertEqual(c.latitudeDegrees(), 5.0)
    XCTAssertEqual(c.longitudeDegrees(), 5.0)
  }
  
  func testBearingTo() {
    let a = Point(latitude: 5, longitude: 5)
    let b = Point(latitude: 5, longitude: 10)
    
    XCTAssertEqual(a.bearingTo(b), 89.781973)
    XCTAssertEqual(a.bearingTo(b, inRadians: true), 1.566991)
  }
  
  func testDestination() {
    let a = Point(latitude: 37.452602, longitude: -122.207069)
    let d = a.destination(308, distance: 50)
    
    XCTAssertEqual(d, Point(latitude: 37.452879, longitude: -122.207515))
  }
  
  func testDistanceTo() {
    let a = Point(latitude: 50.06639, longitude: -5.71472)
    let b = Point(latitude: 58.64389, longitude: -3.07000)
    
    XCTAssertEqualWithAccuracy(a.distanceTo(b), 968853.52, 0.001)
  }
  
  func testIntersectSimple() {
    let a = Point(latitude: 5, longitude: 5)
    let b = Point(latitude: 15, longitude: 15)
    let c = Point(latitude: 5, longitude: 15)
    let d = Point(latitude: 15, longitude: 5)
    var intersection = Point()
    Point.intersectSimple(p: a, p2: b, q: c, q2: d, intersection: &intersection)
    
    XCTAssertEqual(intersection, Point(latitude: 10, longitude: 10))
  }
  
}
